### ISSUE TITLE
Cannot start RD if the home directory is not on the boot drive

### ISSUE THREAD

**Issue by jandubois** [rancher-sandbox/rancher-desktop#253]
User reported this error when the home directory is on a network drive:

```
Error Starting Kubernetes: EXDEV: cross-device link not permittted 
```

**Comment by ericpromislow** [comment]
Weird ... this isn't for the executables that we link into `/usr/local/bin` because we're calling `fs.symlink` which I certainly hope does symlinking and not hardlinking. None of our own code calls `fs.link`, but we have a pile of dependencies that could:

* node_modules/fs-extra/lib/ensure/link.js - called in `CreateLink` and `CreateLinkSync`
  but:  `fsExtra.link` protected by `isUseHardLink` in `builder-util/out/fs.js`
        
* app-builder-lib/out/asar/asarUtil.js - possibly something in the asar pseudo-filesystem
* terser-webpack-plugin/node_modules/cacache/lib/util/move-file.js
* decompress/index.js
* cacache/lib/util/move-file.js

**Comment by jandubois** [comment]
You are right that the symlinks can't be the problem. I think it is the `fs.rename` calls that try to move a file, that are not allowed to cross device boundaries:

```console
src/k8s-engine/hyperkit.ts
243:      await fs.promises.rename(outPath, this.imageFile);

src/k8s-engine/k3sHelper.ts
346:      await fs.promises.rename(workDir, path.join(cacheDir, version));
508:      await fs.promises.rename(workPath, userPath);

src/k8s-engine/wsl.ts
145:      await util.promisify(fs.rename)(outPath, this.distroFile);
```

PS:

> this isn't for the executables that we link into `/usr/local/bin` because we're calling `fs.symlink`

It is irrelevant for this issue, but I don't think these links are made via `fs.symlink` because they have to be done from inside the WSL distro. I think they are created by https://github.com/jandubois/tinyk3s/blob/main/run-k3s#L31-L37

**Comment by ericpromislow** [comment]
Related, but probably not the same:

I copied the *source code* over to an external volume, thinking this might replicate the conditions, but got a different error during startup:

**Error Starting Kubernetes**

```
Command failed: sh -c 'chown root "/Volumes/EXTERNAL/desktop/resources/darwin/docker-machine-driver-hyperkit" 
&& chmod u+s "/Volumes/EXTERNAL/desktop/resources/darwin/docker-machine-driver-hyperkit" 
&& chown :admin "/Volumes/EXTERNAL/desktop/resources/darwin/hyperkit"'
chown: /Volumes/EXTERNAL/desktop/resources/darwin/docker-machine-driver-hyperkit: Operation not permitted
```

**Comment by jandubois** [comment]
> I copied the _source code_ over to an external volume, thinking this might replicate the conditions, but got a different error during startup:

Note that the original issue here is about Windows, not macOS. They use different code paths, so the errors are not comparable.

If you want to use an external volume on macOS, you must make sure you mount it with the `suid` option; otherwise adding the `suid` bit on `docker-machine-driver-hyperkit` will have no effect.

You can do that permanently by adding the `UUID` of your external volume in `/etc/fstab`, like this (from my external ThunderBolt SSD Raid array):

```console
$ cat /etc/fstab
#
# Warning - this file should only be modified with vifs(8)
#
...
# ThunderBlade: mount with suid
UUID=1E6A134F-3306-3560-8D48-56865A4D3889 none hfs rw,suid,auto
```

**Comment by ericpromislow** [comment]
Should that part about mounting it with the `suid` bit be part of the docs, or is this an unlikely path for macos users to follow?

**Comment by jandubois** [comment]
> Should that part about mounting it with the `suid` bit be part of the docs, or is this an unlikely path for macos users to follow?

That was really just for you (or any developer who runs RD from a different device) because you mentioned running from a different device on macOS as a test. As a user you should install the app into `/Applications`, which will be on your boot partition. But I'll create a separate issue for it anyways, just in case.

Please let's keep this Github issue focused on the Windows issue.

**Comment by jandubois** [comment]
@ericpromislow I would recommend against using the shared host directory for this test; it uses a VMware driver and is not a regular disk or network drive, afaict. For example it doesn't seem to map into WSL2.

I would simply allocated another disk to the VM, so you have a `D:` drive as well (that you may need to partition/format).

Also don't just create a "fake" home directory; it also needs to be set in the registry to be entirely consistent.

You can run `lusrmgr.msc` to edit the user properties, and there is a page that allows you change the home directory for a user.

**Comment by jandubois** [comment]
> I would simply allocated another disk to the VM, so you have a `D:` drive as well (that you may need to partition/format).

I have just tried it, and it works as expected:

* power down VM
* Add Device (New Harddisk)
* Boot back up, run "Disk Management", initialize new disk and create "simple volume"
* It becomes drive "E:" because "D:" is my CDROM
* Use `lusrmgr.msc` to change my home directory (on the profile page) to `e:\home\jan`
* Reboot
* Start new Command Prompt and home is now `e:\home\jan`.

**Comment by ericpromislow** [comment]
Finally got a second user running on the VM with a home directory on drive E:, and can't reproduce this problem, although kubernetes fails to start up but `k3s` is running. Error messages, when I get them, vary. One time it reported not being able to connect to a localhost on port 6xxx.

**Comment by jandubois** [comment]
I can confirm that `lusrmgr.msc` is disabled on Windows Home.

I'll try to repro the original issue on a Windows Pro machine now. I suspect that you have to remove all traces of a previous RD install to force RD to download all files again.

**Comment by jandubois** [comment]
It looks like the kubeconfig file is created in the tmp directory and then moved (renamed) to the users home directory. That does not work when the directories are on different drives:

<img width="555" alt="Screen Shot 2021-05-12 at 5 31 03 PM" src="https://user-images.githubusercontent.com/78947/118060673-208c1280-b348-11eb-9dca-901cc3e42bb3.png">

**Comment by ericpromislow** [comment]
I only saw this on the distro file - will look into this as well.

**Comment by jandubois** [comment]
Since the fix didn't make it into `v0.2.0`, here is a workaround by calling the app from the commandline:

```console
set TMP=%HOMEDRIVE%%HOMEPATH%\temp
set TEMP=%TMP%
mkdir %TMP%
"%LOCALAPPDATA%\Programs\Rancher Desktop\Rancher Desktop.exe"
```

**Comment by ericpromislow** [comment]
Verifying that this workaround is necessary.

**Comment by jandubois** [comment]
> Verifying that this workaround is necessary.

@ericpromislow I can assure you that this workaround is necessary with the v0.2.0 release on my setup where the home directory has been moved to a separate drive. 😄

**Comment by jandubois** [comment]
I get a different error with #292 merged:
<img width="549" alt="Screen Shot 2021-05-17 at 9 58 58 PM" src="https://user-images.githubusercontent.com/78947/118592855-810dbc00-b75b-11eb-89f1-563aaf803cc7.png">
So this issue is not yet fixed.

**Comment by mattfarina** [comment]
This was fixed by #304
